modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/outgoing/OutgoingSnapshotsManager.java

```diff
@@ -0,0 +1,145 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.manager.IgniteComponent;
+import org.apache.ignite.internal.table.message.SnapshotMetaRequest;
+import org.apache.ignite.internal.table.message.SnapshotMvDataRequest;
+import org.apache.ignite.internal.table.message.SnapshotRequestMessage;
+import org.apache.ignite.internal.table.message.SnapshotTxDataRequest;
+import org.apache.ignite.internal.table.message.TableMessageGroup;
+import org.apache.ignite.network.MessagingService;
+import org.apache.ignite.network.NetworkAddress;
+import org.apache.ignite.network.NetworkMessage;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Outgoing snapshots manager.
+ */
+public class OutgoingSnapshotsManager implements IgniteComponent {
+    /** Logger. */
+    private static final IgniteLogger LOG = Loggers.forClass(OutgoingSnapshotsManager.class);
+
+    /** Messaging service. */
+    private final MessagingService messagingService;
+
+    /** Map with outgoing snapshots. */
+    private final ConcurrentMap<UUID, OutgoingSnapshot> outgoingSnapshots = new ConcurrentHashMap<>();
+
+    /**
+     * Constructor.
+     *
+     * @param messagingService Messaging service.
+     */
+    public OutgoingSnapshotsManager(MessagingService messagingService) {
+        this.messagingService = messagingService;
+    }
+
+    /**
+     * Returns a messagins service.
+     */
+    public MessagingService messagingService() {
+        return messagingService;
+    }
+
+    @Override
+    public void start() {
+        messagingService.addMessageHandler(TableMessageGroup.class, this::messageHandler);
+    }
+
+    @Override
+    public void stop() throws Exception {
+    }
+
+    /**
+     * Registers an outgoing snapshot in the manager.
+     *
+     * @param snapshotId Snapshot id.
+     * @param outgoingSnapshot Outgoing snapshot.
+     */
+    void startOutgoingSnapshot(UUID snapshotId, OutgoingSnapshot outgoingSnapshot) {
+        outgoingSnapshots.put(snapshotId, outgoingSnapshot);
+    }
+
+    /**
+     * Removes an outgoing snapshot from the manager.
+     *
+     * @param snapshotId Snapshot id.
+     */
+    void stopOutgoingSnapshot(UUID snapshotId) {
+        outgoingSnapshots.remove(snapshotId);
+    }
+
+    private void messageHandler(NetworkMessage networkMessage, NetworkAddress sender, @Nullable Long correlationId) {
+        // Ignore all messages that we can't handle.
+        if (!(networkMessage instanceof SnapshotRequestMessage)) {
+            return;
+        }
+
+        assert correlationId != null;
+
+        OutgoingSnapshot outgoingSnapshot = outgoingSnapshots.get(((SnapshotRequestMessage) networkMessage).id());
+
+        if (outgoingSnapshot == null) {
+            if (LOG.isWarnEnabled()) {
```

Comment:
No need for this `if` here

Reply:
Why?

Reply:
because `Log.warn` will check the same condition internally

Reply:
This applies to all methods, like `info` and `error`, but manual check saves on CPU. It's been shown many times

Reply:
> but manual check saves on CPU. It's been shown many times

Can you please point me to where it is shown?

Reply:
No, it's classified)
I pulled that from my memory. Usually this discussion appears when we talk about `debug` logs, but same reasoning applies to everything else